File: go/24.swap-nodes-in-pairs.go
/*
 * @lc app=leetcode id=24 lang=golang
 *
 * [24] Swap Nodes in Pairs
 */
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	h := head.Next
	first := head
	second := head.Next
	var last *ListNode
	for ; first != nil && second != nil; {
		var nextFirst, nextSecond *ListNode
		nextFirst = second.Next 
		if nextFirst == nil {
			nextSecond = nil
		} else {
			nextSecond = nextFirst.Next
		}

		if last != nil {
			last.Next = second
		}
		last = first
		second.Next = first

		first = nextFirst
		second = nextSecond
	}
	last.Next = first

	return h
}

File: csharp/24.swap-nodes-in-pairs.cs
/*
 * @lc app=leetcode id=24 lang=csharp
 *
 * [24] Swap Nodes in Pairs
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution 
{
    public ListNode SwapPairs(ListNode head) 
    {
        ListNode dummy = new ListNode(int.MaxValue, head);

        ListNode prev = dummy;
        ListNode cur = head;
        ListNode next = null;

        while (cur != null && cur.next != null)
        {
            next = cur.next.next;
            prev.next = cur.next;
            cur.next.next = cur;
            cur.next = next;

            prev = cur;
            cur = next;
        }

        return dummy.next;
    }
}
// @lc code=end

